File: day08/solution.py
import copy
from pathlib import Path
from typing import List, Tuple


CWD = Path(__file__).parent


OPERATORS = {
  "nop": lambda n: lambda index, count: (index+1, count),
  "acc": lambda n: lambda index, count: (index+1, count+n),
  "jmp": lambda n: lambda index, count: (index+n, count),
}


def read_input() -> List[Tuple[str, int]]:
    input_file = CWD.joinpath("input.txt")
    with open(input_file, "r") as reader:
        tmp = [l.split(" ") for l in reader.read().rstrip().split("\n")]
        return [(f, int(arg)) for f, arg in tmp]


def first(code: List[Tuple[str, int]], index: int, counter: int) -> Tuple[int, bool]:
    visited = set()
    while 0 <= index < len(code) and index not in visited:
        visited.add(index)
        op, arg = code[index]
        index, counter = OPERATORS[op](arg)(index, counter)
    return counter, index >= len(code)


def second(code: List[Tuple[str, int]]) -> int:
    visited, index, counter = set(), 0, 0
    while index not in visited or index >= len(code):
        visited.add(index)
        op, arg = code[index]
        res, unlooped = 0, False
        if op != "acc":
            alt_code = copy.copy(code)
            alt_code[index] = ("jmp" if op == "nop" else "nop", code[index][1])
            res, unlooped = first(alt_code, index, counter)
        if unlooped:
            return res
        index, counter = OPERATORS[op](arg)(index, counter)


def run() -> None:
    rules = read_input()
    print("First step:")
    print(first(rules, 0, 0))
    print("\nSecond step:")
    print(second(rules))

File: day08/solution.go
package day8

import (
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
	"time"
)

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

var input = "./input.txt"

type operator func(int, int, int) (int, int)

var operators = map[string]operator{
	"nop": func(x, y, z int) (int, int) { return y + 1, z },
	"acc": func(x, y, z int) (int, int) { return y + 1, z + x },
	"jmp": func(x, y, z int) (int, int) { return x + y, z },
}

// Command in the program
type Command struct {
	typ string
	arg int
}

func (o *Command) isAcc() bool {
	return o.typ == "acc"
}

func (o *Command) invert() {
	switch o.typ {
	case "nop":
		o.typ = "jmp"
	case "jmp":
		o.typ = "nop"
	}
}

func (o *Command) execute(index, count int) (int, int) {
	return operators[o.typ](o.arg, index, count)
}

// ReadInput retrieves the content of the input file
func ReadInput() (res []Command) {
	data, _ := ioutil.ReadFile(input)
	for _, line := range strings.Split(string(data), "\n") {
		substr := strings.Split(line, " ")
		arg, _ := strconv.Atoi(substr[1])
		res = append(res, Command{substr[0], arg})
	}
	return
}

func execute(code []Command, index, count int) (int, error) {
	visited := make(map[int]bool)
	for index >= 0 && index < len(code) {
		if _, ok := visited[index]; ok {
			return count, fmt.Errorf("looped")
		}
		visited[index] = true
		index, count = code[index].execute(index, count)
	}
	return count, nil
}

// Step1 solves step 1
func Step1(code []Command) int {
	defer timeTrack(time.Now(), "Step 1")
	res, _ := execute(code, 0, 0)
	return res
}

func altExecute(code []Command, index, count int) (int, error) {
	code[index].invert()
	return execute(code, index, count)
}

// Step2 solves step 2
func Step2(code []Command) int {
	defer timeTrack(time.Now(), "Step 2")
	index, count := 0, 0
	for index >= 0 && index < len(code) {
		cmd := code[index]
		if !cmd.isAcc() {
			res, err := altExecute(code, index, count)
			if err == nil {
				return res
			}
		}
		index, count = cmd.execute(index, count)
	}
	return -1
}
